superset/db_engine_specs/ocient.py

```diff
@@ -215,32 +222,55 @@ class OcientEngineSpec(BaseEngineSpec):
 
     @classmethod
     def get_table_names(
-        cls, database: "Database", inspector: Inspector, schema: Optional[str]
+        cls, database: Database, inspector: Inspector, schema: Optional[str]
     ) -> List[str]:
         return sorted(inspector.get_table_names(schema))
 
     @classmethod
     def fetch_data(cls, cursor, lim=None):
         rows = super(OcientEngineSpec, cls).fetch_data(cursor)
+                
+        if len(rows) > 0 and type(rows[0]).__name__ == rows:
+            # Peek at the schema to determine which column values, if any,
+            # require sanitization.
+            columns_to_sanitize: List[PlacedSanitizeFunc] = _find_columns_to_sanitize(cursor)
 
-        if (len(rows)) == 0:
-            # No rows were produced
-            return rows
+            if columns_to_sanitize:
+                # At least 1 column has to be sanitized.
+                for row in rows:
+                    for info in columns_to_sanitize:
+                        # Modify the element in-place.
+                        v = row[info.column_index]
+                        row[info.column_index] = info.sanitize_func(v)
 
-        if type(rows[0]).__name__ != 'Row':
-            # TODO what else is returned here?
-            return rows
+        # # We are done with this cursor so we can safely remove it from the cache
+        query_ids_to_delete = [q_id for q_id in cursor_cache if cursor_cache[q_id] == cursor]
+        for q_id in query_ids_to_delete:
+            del cursor_cache[q_id]
+        # # We are done with this cursor so we can safely remove it from the cache
+        query_ids_to_delete = [q_id for q_id in cursor_cache if cursor_cache[q_id] == cursor]
+        for q_id in query_ids_to_delete:
+            del cursor_cache[q_id]
+        return rows
 
-        # Peek at the schema to determine which column values, if any,
-        # require sanitization.
-        columns_to_sanitize: List[PlacedSanitizeFunc] = _find_columns_to_sanitize(cursor)
 
-        if columns_to_sanitize:
-            # At least 1 column has to be sanitized.
-            for row in rows:
-                for info in columns_to_sanitize:
-                    # Modify the element in-place.
-                    v = row[info.column_index]
-                    row[info.column_index] = info.sanitize_func(v)
+    @classmethod
+    def get_cancel_query_id(cls, cursor: Any, query: Query) -> Optional[str]:
+        cursor_cache[query.id] = cursor
+        
+        # Return a Non-None value
+        # If None is returned, Superset will not call cancel_query
+        return 'DUMMY_VALUE'
 
-        return rows
+    @classmethod
+    def cancel_query(cls, cursor: Any, query: Query, cancel_query_id: str) -> bool:
+
+        if query.id in cursor_cache:
+            cursor.execute(f'CANCEL {cursor_cache[query.id].query_id}')
+            # Query has been cancelled, so we can safely remove the cursor from the cache
+            del cursor_cache[query.id]
```

Comment:
I don't see how this works actually, the key type is not an id?
```
cursor_cache: Dict[Query, pyocient.Cursor]= dict()
```

Reply:
oh wait, I see we do insert a query id in `fetch_data()`. Maybe the typing syntax is supposed to be UUID?

Reply:
makes more sense now, totally possible that the `Query` type _is_ a unique identifier itself. I just assumed that the `Query` object contained more than just an identifier